### PR TITLE
added if_exists with tests

### PR DESCRIPTION
**Pull request by fnc12**

### CODE REVIEW

File: dev/storage_base.h [hunk 1]
```diff
@@ -885,15 +924,40 @@ namespace sqlite_orm {
                 return result;
             }
 
-            void drop_table_internal(sqlite3* db, const std::string& tableName) {
+            void drop_table_internal(sqlite3* db, const std::string& tableName, bool ifExists) {
                 std::stringstream ss;
-                ss << "DROP TABLE " << streaming_identifier(tableName) << std::flush;
+                ss << "DROP TABLE";
```
Comment (trueqbit) [inline dev/storage_base.h:929]: The trailing whitespace character is necessary.

Comment (trueqbit) [inline dev/storage_base.h:929]: The trailing whitespace character is necessary.

File: dev/storage_base.h [hunk 2]
```diff
@@ -885,15 +924,40 @@ namespace sqlite_orm {
...
+                ss << streaming_identifier(tableName) << std::flush;
                 perform_void_exec(db, ss.str());
             }
 
-            static int collate_callback(void* arg, int leftLen, const void* lhs, int rightLen, const void* rhs) {
-                auto& f = *(collating_function*)arg;
-                return f(leftLen, lhs, rightLen, rhs);
+            void drop_index_internal(const std::string& indexName, bool ifExists) {
+                std::stringstream ss;
+                ss << "DROP INDEX";
```
Comment (trueqbit) [inline dev/storage_base.h:939]: The trailing whitespace character is necessary.

File: dev/storage_base.h [hunk 3]
```diff
@@ -885,15 +924,40 @@ namespace sqlite_orm {
...
+                if(ifExists) {
+                    ss << " IF EXISTS";
```
Comment (trueqbit) [inline dev/storage_base.h:941]: A trailing instead of a leading whitespace character is necessary.

File: dev/storage_base.h [hunk 4]
```diff
@@ -885,15 +924,40 @@ namespace sqlite_orm {
...
+                }
+                ss << quote_identifier(indexName) << std::flush;
+                perform_void_exec(this->get_connection().get(), ss.str());
+            }
+
+            void drop_trigger_internal(const std::string& triggerName, bool ifExists) {
+                std::stringstream ss;
+                ss << "DROP TRIGGER";
```
Comment (trueqbit) [inline dev/storage_base.h:949]: The trailing whitespace character is necessary.

File: dev/storage_base.h [hunk 5]
```diff
@@ -885,15 +924,40 @@ namespace sqlite_orm {
...
+                if(ifExists) {
+                    ss << " IF EXISTS";
```
Comment (trueqbit) [inline dev/storage_base.h:951]: A trailing instead of a leading whitespace character is necessary.

File: dev/storage_base.h [hunk 6]
```diff
@@ -885,15 +924,40 @@ namespace sqlite_orm {
                 return result;
             }
 
-            void drop_table_internal(sqlite3* db, const std::string& tableName) {
+            void drop_table_internal(sqlite3* db, const std::string& tableName, bool ifExists) {
                 std::stringstream ss;
-                ss << "DROP TABLE " << streaming_identifier(tableName) << std::flush;
+                ss << "DROP TABLE";
+                if(ifExists) {
+                    ss << " IF EXISTS";
```
Comment (trueqbit) [inline dev/storage_base.h:931]: A trailing instead of a leading whitespace character is necessary.

**Review by trueqbit — CHANGES_REQUESTED** [review]

File: dev/storage_base.h (hunk 1 above)
Comment (fnc12) [reply dev/storage_base.h:929]: done

File: dev/storage_base.h (hunk 2 above)
Comment (fnc12) [reply dev/storage_base.h:939]: done

File: dev/storage_base.h (hunk 3 above)
Comment (fnc12) [reply dev/storage_base.h:941]: done

File: dev/storage_base.h (hunk 4 above)
Comment (fnc12) [reply dev/storage_base.h:949]: done

File: dev/storage_base.h (hunk 5 above)
Comment (fnc12) [reply dev/storage_base.h:951]: done

File: dev/storage_base.h (hunk 6 above)
Comment (fnc12) [reply dev/storage_base.h:931]: done

**Comment by fnc12** [comment]
@trueqbit I fixed your concerns with a different way. I add one whitespace before every dedicated AST portion during serialization. This is the unique way which works everywhere

**Pull request merged**